alg/teca_temporal_reduction.cxx

```diff
@@ -0,0 +1,1083 @@
+            req_arrays.insert(vv_mask);
+
+            // get the fill value
+            teca_metadata array_atts;
+            atts.get(array, array_atts);
+            if (this->fill_value != -1)
+            {
+                fill_value = this->fill_value;
+                std::cerr << "this->fill_value = " << fill_value << std::endl;
+            }
+            else if (array_atts.has("_FillValue"))
+            {
+                array_atts.get("_FillValue", fill_value);
+                std::cerr << "_FillValue = " << fill_value << std::endl;
+            }
+            else if (array_atts.has("missing_value"))
+            {
+                array_atts.get("missing_value", fill_value);
+                std::cerr << "missing_value = " << fill_value << std::endl;
+            }
+        }
+
+        // create and initialize the operator
+        p_reduction_operator op = reduction_operator_factory::New(this->operator_name);
+        op->initialize(fill_value);
+
+        // save the operator
+        this->op[array] = op;
+    }
+
+    // generate one request for each time step in the interval
+    std::vector<teca_metadata> up_reqs;
+    std::string request_key;
+    md.get("index_request_key", request_key);
+    unsigned long req_id;
+    req_in.get(request_key, req_id);
+    int i = this->indices[req_id].start_index;
+    while (i <= this->indices[req_id].end_index)
+    {
+        //std::cerr << i << " " << this->indices[req_id].end_index << std::endl;
+        teca_metadata req(req_in);
+        req.set("arrays", req_arrays);
+        req.set(request_key, i);
+        up_reqs.push_back(req);
+        i += 1;
+    }
+
+    return up_reqs;
+}
+
+// --------------------------------------------------------------------------
+const_p_teca_dataset teca_temporal_reduction::execute(
+    unsigned int port,
+    const std::vector<const_p_teca_dataset> &data_in,
+    const teca_metadata &req_in,
+    int streaming)
+{
+    (void)port;
+
+    // get the requested ind
+    std::string request_key;
+    req_in.get("index_request_key", request_key);
+    unsigned long req_id;
+    req_in.get(request_key, req_id);
+
+    int device_id = -1;
+    req_in.get("device_id", device_id);
+
+#ifdef TECA_DEBUG
+    std::cerr << teca_parallel_id()
+        << "teca_temporal_reduction::execute request "
+        << req_id << " device " << device_id << " (" << this->indices[req_id].start_index
+        << " - " << this->indices[req_id].end_index
+        << "), reducing " << data_in.size() << ", "
+        << streaming << " remain" << std::endl;
+#endif
+
+    // copy the first mesh
+    const_p_teca_cartesian_mesh mesh_in
+        = std::dynamic_pointer_cast<const teca_cartesian_mesh>(data_in[0]);
+    p_teca_cartesian_mesh mesh_out = teca_cartesian_mesh::New();
+    mesh_out->shallow_copy(
+        std::const_pointer_cast<teca_cartesian_mesh>(mesh_in));
+    p_teca_array_collection arrays_out = mesh_out->get_point_arrays();
```

Comment:
It's convenient to shallow copy the input mesh to grab metadata etc, but you'll need to allocate new arrays on the assigned device for the output. if you don't do this, a couple of bad things can happen. 1) you'll be modifying the input data, which is against the rules of the system 2) in the case the input was allocated on the CPU and you are trying to compute on the GPU you'll end up operating on a temporary and simply have the original (the one that was shallow copied) array at the end.